hera_qm/xrfi.py

```diff
@@ -607,11 +685,215 @@ def xrfi_h1c_pipe(uv, Kt=8, Kf=8, sig_init=6., sig_adj=2., px_threshold=0.2,
+
+    # Initial run on cal data products
+    alg = 'detrend_medfilt'
+    # Calculate metric on abscal data
+    uvc_a = UVCal()
+    uvc_a.read_calfits(acalfits_file)
+    uvf_apriori = UVFlag(uvc_a, mode='flag', copy_flags=True)
+    uvf_ag, uvf_agf = xrfi_pipe(uvc_a, alg=alg, Kt=kt_size, Kf=kf_size, xants=xants,
+                                cal_mode='gain', sig_init=sig_init, sig_adj=sig_adj)
+    uvf_ax, uvf_axf = xrfi_pipe(uvc_a, alg=alg, Kt=kt_size, Kf=kf_size, xants=xants,
+                                cal_mode='tot_chisq', sig_init=sig_init, sig_adj=sig_adj)
+
+    # Calculate metric on omnical data
+    uvc_o = UVCal()
+    uvc_o.read_calfits(ocalfits_file)
+    flag_apply(uvf_apriori, uvc_o, keep_existing=True)
+    uvf_og, uvf_ogf = xrfi_pipe(uvc_o, alg=alg, Kt=kt_size, Kf=kf_size, xants=xants,
+                                cal_mode='gain', sig_init=sig_init, sig_adj=sig_adj)
+    uvf_ox, uvf_oxf = xrfi_pipe(uvc_o, alg=alg, Kt=kt_size, Kf=kf_size, xants=xants,
+                                cal_mode='tot_chisq', sig_init=sig_init, sig_adj=sig_adj)
+
+    # Calculate metric on model vis
+    uv_v = UVData()
+    uv_v.read(model_file)
+    uvf_v, uvf_vf = xrfi_pipe(uv_v, alg=alg, xants=[], Kt=kt_size, Kf=kf_size,
+                              sig_init=sig_init, sig_adj=sig_adj)
+
+    # Combine the metrics together
+    uvf_metrics = uvf_v.combine_metrics([uvf_og, uvf_ox, uvf_ag, uvf_ax],
+                                        method='quadmean', inplace=False)
+    alg_func = algorithm_dict[alg]
+    uvf_metrics.metric_array[:, :, 0] = alg_func(uvf_metrics.metric_array[:, :, 0],
+                                                 flags=~uvf_metrics.weights_array[:, :, 0].astype(np.bool),
+                                                 Kt=kt_size, Kf=kf_size)
+
+    # Flag on combined metrics
+    uvf_f = flag(uvf_metrics, nsig_p=sig_init)
+    uvf_fws = watershed_flag(uvf_metrics, uvf_f, nsig_p=sig_adj, inplace=False)
+    # OR everything together for initial flags
+    uvf_apriori.to_waterfall(method='and', keep_pol=False)
+    uvf_init = uvf_fws | uvf_ogf | uvf_oxf | uvf_agf | uvf_axf | uvf_vf | uvf_apriori
+
+    # Write out initial (combined) metrics and flags
+    basename = qm_utils.strip_extension(os.path.basename(data_file))
+    outfile = '.'.join([basename, init_metrics_ext])
+    outpath = os.path.join(dirname, outfile)
+    uvf_metrics.write(outpath, clobber=True)
+    outfile = '.'.join([basename, init_flags_ext])
+    outpath = os.path.join(dirname, outfile)
+    uvf_init.write(outpath, clobber=True)
+
+    # Second round -- use init flags to mask and recalculate everything
+    # Read in data file
+    uv_d = UVData()
+    uv_d.read(data_file)
+    for uv in [uvc_o, uvc_a, uv_v, uv_d]:
+        flag_apply(uvf_init, uv, keep_existing=True, force_pol=True)
+
+    # Do next round of metrics
+    alg = 'detrend_meanfilt'  # Change to meanfilt because it can mask flagged pixels
```

Comment:
Same comment as before about `alg`

Also, given the amount of repeated code in this function, would it make sense to make a helper function that does omni/abs/omni model flags all together?

Reply:
I thought about that, but the fact that we have one extra (data), and it's only repeated twice, I didn't think it was worth it.